### PR TITLE
posm: CLEAR_OR_TAKE

### LINKED ISSUE #248: posm: CLEAR_OR_TAKE
Add a command to posm "CLEAR_OR_TAKE" which clears up to a certain limit otherwise it will take

_Originally posted by @saucepoint in https://github.com/Uniswap/v4-periphery/pull/243#discussion_r1701901505_

### PR DESCRIPTION
**Pull request by saucepoint**
## Related Issue
Closes #248 

## Description of changes

Refactor CLEAR to use `_take()` if the delta exceeds the user's limit

### CODE REVIEW

File: src/PositionManager.sol [hunk 1]
```diff
@@ -208,11 +209,18 @@ contract PositionManager is
...
     /// @dev integrators may elect to forfeit positive deltas with clear
-    /// provides a safety check that amount-to-clear is less than a user-provided maximum
-    function _clear(Currency currency, uint256 amountMax) internal {
+    /// if the forfeit amount exceeds the user-specified max, the amount is taken instead
+    function _clearOrTake(Currency currency, uint256 amountMax) internal {
         int256 currencyDelta = poolManager.currencyDelta(address(this), currency);
-        if (uint256(currencyDelta) > amountMax) revert ClearExceedsMaxAmount(currency, currencyDelta, amountMax);
-        poolManager.clear(currency, uint256(currencyDelta));
+        if (currencyDelta < 0) revert CannotClearNegativeDelta(currency);
+        uint256 delta = uint256(currencyDelta); // safe since we know its positive
```
Comment (saucepoint) [inline src/PositionManager.sol:216]: wonder if this should be replaced with `_getFullTakeAmount`? the error might need to be updated

```solidity
    function _getFullTakeAmount(Currency currency) internal view returns (uint256 amount) {
        int256 _amount = poolManager.currencyDelta(address(this), currency);
        // If the amount is negative, it should be settled not taken.
        if (_amount < 0) revert IncorrectUseOfTake();
        amount = uint256(_amount);
    }
```

Comment (snreynolds) [reply src/PositionManager.sol:216]: Yeah we could change both of those functions to be instead 

_getFullNegativeDelta and _getFullPositiveDelta and the error be more general like DeltaIsNotNegative()/DeltaIsNotPositive()

**Review by snreynolds — COMMENTED** [review]
this looks great but I know there are merge conflicts cause we just merged something for router

File: src/base/DeltaResolver.sol [hunk 2]
```diff
@@ -46,17 +46,17 @@ abstract contract DeltaResolver is ImmutableState {
...
+    function _getFullPositiveDelta(Currency currency) internal view returns (uint256 amount) {
         int256 _amount = poolManager.currencyDelta(address(this), currency);
         // If the amount is positive, it should be taken not settled for.
-        if (_amount > 0) revert IncorrectUseOfSettle();
+        if (_amount > 0) revert DeltaNotPositive(currency);
         amount = uint256(-_amount);
     }
 
-    function _getFullTakeAmount(Currency currency) internal view returns (uint256 amount) {
+    function _getFullNegativeDelta(Currency currency) internal view returns (uint256 amount) {
```
Comment (hensha256) [inline src/base/DeltaResolver.sol:56]: yeah nice renamed

File: src/PositionManager.sol [hunk 3]
```diff
@@ -251,18 +252,29 @@ contract PositionManager is
...
+        // forfeit the delta if its less than or equal to the user-specified limit
+        if (delta <= amountMax) {
+            poolManager.clear(currency, delta);
+        } else {
+            _take(currency, msgSender(), delta);
+        }
+    }
+
+    /// @dev uses this addresses balance to settle a negative delta
+    function _settleWithBalance(Currency currency) internal {
```
Comment (snreynolds) [inline src/PositionManager.sol:268]: remove this

File: src/PositionManager.sol [hunk 4]
```diff
@@ -251,18 +252,29 @@ contract PositionManager is
...
+        // set the payer to this address, performs a transfer.
+        _settle(currency, address(this), _getFullDebit(currency));
     }
 
     function _settlePair(Currency currency0, Currency currency1) internal {
         // the locker is the payer when settling
         address caller = msgSender();
-        _settle(currency0, caller, _getFullSettleAmount(currency0));
-        _settle(currency1, caller, _getFullSettleAmount(currency1));
+        _settle(currency0, caller, _getFullDebit(currency0));
```
Comment (snreynolds) [inline src/PositionManager.sol:276]: debt

File: src/base/DeltaResolver.sol (hunk 2 above)
Comment (snreynolds) [reply src/base/DeltaResolver.sol]: we did a bettttter rename i promise

**Review by snreynolds — APPROVED** [review]

**Pull request merged**